File: loaders/PyMarkLoader.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.DataInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

@SuppressWarnings("rawtypes")
public class PyMarkLoader {

	private HashMap<String, Object> modules;
	public HashMap<String, Object> getModules() { return modules; }
	
	private ArrayList<String> moduleNames;
	public ArrayList<String> getModuleNames(){ return moduleNames; }
	
	private String path;
	public String getPath(){ return path; }
	
    public final static short LIST 		   = 1;
    public final static short LONG_LIST    = 2;
    public final static short DICT         = 3;
    public final static short LONG_DICT    = 4;
    public final static short TUPLE        = 5;
    public final static short LONG_TUPLE   = 6;
    public final static short STRING       = 7;
    public final static short LONG_STRING  = 8;
    public final static short INT          = 9;
    public final static short LONG         = 10;
    public final static short FLOAT        = 11;
    public final static short DOUBLE       = 12;
    public final static short REFERENCE    = 13;
    public final static short NONE         = 14;	
	
    private final String REF_PREFIX = "*PYMARK-REFERENCE*";
    
	@SuppressWarnings("unchecked")
	public PyMarkLoader(String _path)
	{
		path = _path;
		modules = new HashMap();
		moduleNames = new ArrayList<String>();
		
		LoadModuleNames();
		
		ClearLog();
		
		LogMsg("PyMarkLoader Constructed");
	}
	
	private void LogMsg(String msg)
	{
	    try {

		      String logLoc = path+"\\Logs\\java_loader.log";
		    	
		      BufferedWriter out = new BufferedWriter(new FileWriter(logLoc, true)); 
		      out.write(msg+"\n");
		      out.close();
		    } catch (IOException e) {
		      e.printStackTrace();
		    }
	}
	
	private void ClearLog()
	{
	    try {

		      String logLoc = path+"\\Logs\\java_loader.log";
		    	
		      BufferedWriter out = new BufferedWriter(new FileWriter(logLoc, false)); 
		      out.write("");
		      out.close();
		    } catch (IOException e) {
		      e.printStackTrace();
		    }
	}
	
	
    private boolean procDone(Process p) {
        try {
            @SuppressWarnings("unused")
			int v = p.exitValue();
            return true;
        }
        catch(IllegalThreadStateException e) {
            return false;
        }
    }	
	
	public void Compile()
	{	
		try{
		
		LogMsg("Compiling all modules...");
		
		String command = "python \""+path+"\\PyMark.py\"";
		
        Process p = Runtime.getRuntime().exec(command);

        // read the standard output of the command

        BufferedReader stdInput = new BufferedReader(new InputStreamReader(p.getInputStream()));

        String output = "";
		while (!procDone(p)) {
			String s;
			while((s = stdInput.readLine()) !=null){
				output += (s+"\n");
			}
		}
		stdInput.close();
		
        LogMsg("\n>>>>>");
        LogMsg(output);
        LogMsg(">>>>>\n");
		
		}
		catch(Exception e)
		{
			LogMsg("ERROR: Could not compile modules: "+e.toString());
		}
	}
	
	public void CompileModules(List<String> names,String flags)
	{
		try{
			
			String moduleNames = " ";
			for (int i=0;i<names.size();i++)
			{
				moduleNames += (names.get(i) + " ");
			}
			
			LogMsg("Compiling modules"+moduleNames);			
			
			String command = "python \""+path+"\""+moduleNames+" "+flags;
			
	        Process p = Runtime.getRuntime().exec(command);

	        // read the standard output of the command

	        BufferedReader stdInput = new BufferedReader(new InputStreamReader(p.getInputStream()));

	        String output = "";
			while (!procDone(p)) {
				String s;
				while((s = stdInput.readLine()) !=null){
					output += (s+"\n");
				}
			}
			stdInput.close();
			
	        LogMsg("\n>>>>>");
	        LogMsg(output);
	        LogMsg(">>>>>\n");
			
			}
			catch(Exception e)
			{
				LogMsg("ERROR: Could not compile modules: "+e.toString());
			}		
	}
	
	private void LoadModuleNames()
	{
		moduleNames.clear();
		
		String directory = path+"\\Compiled";
		File dir = new File(directory);
		String[] children = dir.list();
		
		if (children == null) {} 
		else
		{ 
			for (int i=0; i<children.length; i++)
			{
				String name = children[i];
				if (name.endsWith(".pm"))
					moduleNames.add(name.substring(0, name.length()-3));
			}
		}
	}
	
	public void Load()
	{
		LogMsg("Loading all modules...");
		LoadModuleNames();
		
		LoadModules(moduleNames);
	}
	public void LoadModules(ArrayList<String> names)
	{
		for (int i=0;i<names.size();i++)
		{
			try{
			
			String name = names.get(i);
			
			LogMsg("Loading module "+name+"...");
			
			String file = path+"\\Compiled\\"+name+".pm";
			FileInputStream fstream = new FileInputStream(file);
			DataInputStream stream = new DataInputStream(fstream);
			
			modules.put(name, BuildObject(stream));
			}
			catch(Exception e)
			{
				String name = names.get(i);
				LogMsg("ERROR: Could not loade module \""+name+"\" :"+e.toString());
			}
		}
	}
	
	@SuppressWarnings("unchecked")
	public Object BuildObject(DataInputStream stream) throws Exception
	{
		short type = readShort(stream);
		
		if ((type == LIST) || (type == TUPLE))
		{
			ArrayList retObject = new ArrayList();
			int len = readInt(stream);
			
			for (int i = 0;i<len;i++)
			{
				retObject.add(BuildObject(stream));
			}
			return retObject;
		}
		else if ((type == LONG_LIST) || (type == LONG_TUPLE))
		{
			ArrayList retObject = new ArrayList();
			long len = readLong(stream);
			
			for (long i = 0;i<len;i++)
			{
				retObject.add(BuildObject(stream));
			}
			return retObject;
		}
		else if (type == DICT)
		{
			HashMap retObject = new HashMap();
			int len = readInt(stream);
			
			for (int i=0;i<len;i++)
			{
				ArrayList keypair = (ArrayList)BuildObject(stream);
				retObject.put(keypair.get(0), keypair.get(1));
			}
			
			return retObject;
		}
		else if (type == LONG_DICT)
		{
			HashMap retObject = new HashMap();
			long len = readLong(stream);
			
			for (long i=0;i<len;i++)
			{
				ArrayList keypair = (ArrayList)BuildObject(stream);
				retObject.put(keypair.get(0), keypair.get(1));
			}
			
			return retObject;
		}
		else if (type == STRING)
		{
			return readString(stream);
		}
		else if (type == LONG_STRING)
		{
			return readLongString(stream);
		}
		else if (type == INT)
		{
			return readInt(stream);
		}
		else if(type == LONG)
		{
			return readLong(stream);
		}
		else if (type == FLOAT)
		{
			return readFloat(stream);
		}
		else if (type == DOUBLE)
		{
			return readDouble(stream);
		}
		else if (type == REFERENCE)
		{
			ArrayList retObject = new ArrayList();
			int len = readInt(stream);
			
			retObject.add(REF_PREFIX);
			for (int i = 0;i<len;i++)
			{
				retObject.add(BuildObject(stream));
			}
			return retObject;
		}
		else if (type == NONE)
		{
			return null;
		}
		else
		{
			throw new Exception("ERROR: Unknown type index: "+((Short)type).toString()+" perhaps a badly formed .pm file?");
		}
		
	}
	
	private double readDouble(DataInputStream stream) throws IOException
	{
		double p1 = stream.readDouble();
		return p1;
	}
	
	public void BuildReferences()
	{
        for ( String key : modules.keySet() )
        {
        	modules.put(key, FollowReferences(modules.get(key)));
        }
	}
	
	@SuppressWarnings("unchecked")
	public Object FollowReferences(Object o)
	{
		if (o == null)
		{
			return null;
		}
		if (o.getClass() == ArrayList.class) 
		{		
			/* if is reference */
			if (((ArrayList) o).get(0) == REF_PREFIX)
			{
				((ArrayList) o).remove(0); // Pop the reference off the list.
				o = Get((ArrayList) o,modules);
			}
			else
			{
				/* Follow all references for all of its objects */
				for(int i=0;i<((ArrayList)o).size();i++)
				{
					((ArrayList)o).set(i, FollowReferences(((ArrayList) o).get(i)) );
				}
			}
		}
		else if (o.getClass() == HashMap.class) 
		{
			/* Follow all references for all of its objects */
			for ( Object key : ((HashMap) o).keySet() )
			{
				 ((HashMap) o).put(key, FollowReferences(((HashMap) o).get(key)));
			}
		}
		
		return o;
	}
	
	public Object Get(String reference)
	{
		return Get(reference,modules);
	}
	
	@SuppressWarnings("unchecked")
	public Object Get(String reference,Object domain)
	{
		ArrayList linkList = new ArrayList();
		
		String[] splits = reference.split("\\.");
		for(int i=0;i<splits.length;i++)
		{	
			String s = splits[i];
			
			try{
				Integer.parseInt(s);
				linkList.add(Integer.parseInt(s));
			}
			catch (Exception e)
			{
				linkList.add(s);
			}
		}
		//Collections.reverse(linkList);
		
		return Get(linkList,domain);
	}
	
	public Object Get(ArrayList linkList,Object domain)
	{
		Object nextKey = linkList.remove(0);
		Object nextObject = null;
		
		if (domain.getClass() == HashMap.class)
		{
			nextObject = ((HashMap)domain).get(nextKey); 
		}
		else if (domain.getClass() == ArrayList.class)
		{
			nextObject = ((ArrayList)domain).get((Integer)nextKey);
		}
		
		if (linkList.size() == 0)
		{
			return nextObject;
		}
		else
		{
			return Get(linkList,nextObject);
		}
	}
	
	
	/* Functions for reading in values */
	
	private short readShort(DataInputStream stream) throws IOException
	{
		return (short)stream.read();
	}
	
	private int readInt(DataInputStream stream) throws IOException
	{
		short p1 = (short)stream.read();
		short p2 = (short)stream.read();
		
		return (p1 << 8) + p2;
	}
	
	private long readLong(DataInputStream stream) throws IOException
	{
		short p1 = (short)stream.read();
		short p2 = (short)stream.read();
		short p3 = (short)stream.read();
		short p4 = (short)stream.read();
		
		return (p1 << 24) + (p2 << 16) + (p3 << 8) + (p4);
	}
	
	private String readString(DataInputStream stream) throws IOException
	{
		int len = readInt(stream);
		String retString = "";
		for (int i = 0;i<len;i++)
		{
			retString += (char)stream.read();
		}
		return retString;
	}
	
	private String readLongString(DataInputStream stream) throws IOException
	{
		long len = readInt(stream);
		String retString = "";
		for (long i = 0;i<len;i++)
		{
			retString += (char)stream.read();
		}
		return retString;
	}
	
	private float readFloat(DataInputStream stream) throws IOException
	{
		float p1 = stream.readFloat();
		return p1;
	}	
}

File: loaders/PyMarkLoader.cpp
#include "StdAfx.h"
#include "PyMarkLoader.hpp"

/*

PyMark Loader class.

So there are a couple of things to note about the C++ implementation. Namely some of the limitations.

	1. No Load() function or ModuleNames list.
		
		The only reason for this is that there isn't a standard way in C++ to list directory contents (although boost is becoming standard enough).
		So I decided not to implement it one way or the other. I'll let the user do this with whatever filesystem library they are using.
		There is just a LoadModules() function which takes a list of strings. And this should work fine.

	2. No BuildReferences().
		
		Because in this implementation we're throwing around void* - there isn't a way to tell if this is a List or a Dict or what.
		So it isn't possible to traverse recurrsively looking for references.
		Instead references are stored in their string form, and it's up to the user to use the Get() function on these when its needed.

	3. Dictionaries keys are strings only.
		
		Again, because we can't get the type from a void*, the only way to know how to look up a reference is if the current key in the chain is an "int" or a "string".
		Also C++ doesn't supply the generalized mapping of any object as in some other languages.

Other than that it should be your implementation as standard.
There are a couple of "System" objects which I'll try to remove in time and replace with only std stuff.

*/

/* Constructors and Destructors */

PyMarkLoader::PyMarkLoader(string _path)
{
	path = new string(_path);
	modules = new Dict;

	ClearLog();

	LogMsg("PyMarkLoader constructed");
}

PyMarkLoader::~PyMarkLoader()
{
	delete path;
	delete modules;
}

/* Getters and Setters */

Dict* PyMarkLoader::GetModules()
{
	return modules;
}

string* PyMarkLoader::GetPath()
{
	return path;
}

/* Methods */

#define LOGPATH "\\Logs\\Cpp_loader.log"

void PyMarkLoader::LogMsg(string msg)
{
	string logPath = (*path)+LOGPATH;

	ofstream logFile;
	logFile.open(logPath.c_str(),ios::app);
	logFile << msg+"\n";
	logFile.close();
}

void PyMarkLoader::ClearLog(void)
{
	string logPath = (*path)+LOGPATH;

	ofstream logFile;
	logFile.open(logPath.c_str());
	logFile << "";
	logFile.close();
}

void PyMarkLoader::Compile(void)
{
	LogMsg("Compiling all modules...");

	char   psBuffer[128];
	FILE   *pPipe;

	string command =  "python \""+(*path)+"\\PyMark.py\"";

	if( (pPipe = _popen( command.c_str(), "r" )) == NULL )
	  exit( 1 );

	/* Read pipe until end of file. */
	string result = "";

	while( !feof( pPipe ) )
	{
	  if( fgets( psBuffer, 128, pPipe ) != NULL )
		 result += psBuffer;
	}
	
	LogMsg("\n>>>>>");

	LogMsg(result);

	LogMsg(">>>>>\n");
}

void PyMarkLoader::CompileModules(vector<string> names,string flags)
{
	string namesList = "";
	
	for (unsigned int i=0;i<names.size();i++)
	{
		namesList += (names[i] + " ");
	}

	LogMsg("Compiling modules "+namesList);

	char   psBuffer[128];
	FILE   *pPipe;

	string command =  "python \""+(*path)+"\\PyMark.py\" "+namesList+" "+flags;

	if( (pPipe = _popen( command.c_str(), "r" )) == NULL )
	  exit( 1 );

	/* Read pipe until end of file. */
	string result = "";

	while( !feof( pPipe ) )
	{
	  if( fgets( psBuffer, 128, pPipe ) != NULL )
		 result += psBuffer;
	}
	
	LogMsg("\n>>>>>");

	LogMsg(result);

	LogMsg(">>>>>\n");
}

void PyMarkLoader::LoadModules(vector<string> names)
{
	for (unsigned int i=0;i<names.size();i++)
	{
		string* name = new string(names[i]);
		
		LogMsg("Loading Module \""+(*name)+"\"...");
		
		string file_path = (*path)+"\\Compiled\\"+(*name)+".pm";

		ifstream stream;
		stream.open(file_path.c_str(), ios::in | ios::binary);

		modules->operator [](name->c_str()) = BuildObject(&stream);
		
		stream.close();
	}
}

void* PyMarkLoader::BuildObject(ifstream* stream)
{
	short type = ReadShort(stream);

	if (type == NONE)
	{
		return NULL;
	}
	else if ((type == LIST) || (type == TUPLE))
	{
		List* ret = new List(); 
		
		int len = ReadInt(stream);
		for (int i=0;i<len;i++)
		{
			ret->push_back(BuildObject(stream));
		}
		return ret;
	}
	else if ((type == LONG_LIST) || (type == LONG_TUPLE))
	{
		List* ret = new List(); 
		
		long len = ReadLong(stream);
		for (long i=0;i<len;i++)
		{
			ret->push_back(BuildObject(stream));
		}
		return ret;
	}
	else if (type == DICT)
	{
		/* For now lets just assume the hash is a string. We can add an assertion later. Otherwise things just get too complicated in C++ >_< */
		Dict* ret = new Dict();

		int len = ReadInt(stream);
		for (int i=0;i<len;i++)
		{
			short tupleType = ReadShort(stream); // Ignore the type info. We know it is a tuple.
			int tupleLength = ReadInt(stream); // Ignore the length, we know its 2;
			short stringType = ReadShort(stream); // Ingore again, we know its a string.

			string* key = new string(ReadString(stream));
			void* obj = BuildObject(stream);

			ret->operator [](key->c_str()) = obj;
		}

		return ret;
	}
	else if (type == LONG_DICT)
	{
		/* For now lets just assume the hash is a string. We can add an assertion later. Otherwise things just get too complicated in C++ >_< */
		Dict* ret = new Dict();

		long len = ReadInt(stream);
		for (long i=0;i<len;i++)
		{
			short tupleType = ReadShort(stream); // Ignore the type info. We know it is a tuple.
			int tupleLength = ReadInt(stream); // Ignore the length, we know its 2;
			short stringType = ReadShort(stream); // Ingore again, we know its a string.

			string* key = new string(ReadString(stream));
			void* obj = BuildObject(stream);
			ret->operator [](key->c_str()) = obj;
		}

		return ret;
	}
	else if (type == INT)
	{
		int* val = new int(ReadInt(stream));
		return val;
	}
	else if (type == LONG)
	{
		long* val = new long(ReadLong(stream));
		return val;
	}
	else if (type == FLOAT)
	{
		float* val = new float(ReadFloat(stream));
		return val;
	}
	else if (type == DOUBLE)
	{
		double* val = new double(ReadDouble(stream));
		return val;
	}
	else if (type == STRING)
	{
		string* val = new string(ReadString(stream));
		return val;
	}
	else if (type == LONG_STRING)
	{
		string* val = new string(ReadLongString(stream));
		return val;
	}
	else if (type == REFERENCE)
	{
		string* ref = new string("");

		int len = ReadInt(stream);
		for (int i=0;i<len;i++)
		{
			short type = ReadShort(stream);
			if (type == STRING)
			{
				string val = string(ReadString(stream));
				ref->operator +=(val.c_str());
			}
			else if (type == INT)
			{
				int i = ReadInt(stream);
				std::string s;
				std::stringstream out;
				out << i;
				s = out.str();
				ref->operator +=(s.c_str());
			}
			
			if (i!=(len-1))
			{
				ref->operator +=(".");
			}
		}
		return ref;
	}
	else
	{
		throw "ERROR: Unknown type index \""+type.ToString()+"\", perhaps a badly formed .pm file?";
		return NULL;
	}

}


void* PyMarkLoader::Get(string reference)
{
	return Get(reference,modules);
}

void* PyMarkLoader::Get(string reference,void* domain)
{
	List linkList = List();
	string* buffer = new string();
	for (unsigned int i=0;i<reference.length();i++)
	{
		char c = reference[i];
		if (c == '.')
		{
			string* add = new string(*buffer);
			linkList.push_back(add);
			buffer->clear();
		}
		else
		{
			buffer->operator +=(c);
		}
	}
	string* add = new string(*buffer);
	linkList.push_back(add); // String wont end in a . so pushing back the buffer again.

	delete buffer; // aaannndd we shouldn't need this anymore.

	return Get(linkList,domain);
}

void* PyMarkLoader::Get(List linkList,void* domain)
{
	/* Because in C++ we've decided dictionaries can only be referenced with strings. We just test if this is a string or if we should cast to int */
	string* nextKeyPtr = (string*)linkList[0];

	int index = 0;
	bool isList = true;
	
	System::String^ testString = gcnew System::String(nextKeyPtr->c_str());
	try
	{
		index = int::Parse(testString);
	}
	catch (...)
	{
		isList = false;
	}
	delete testString;

	void* nextObject;
	if (isList)
	{
		nextObject = ((List*)domain)->operator [](index);
	}
	else
	{
		nextObject = ((Dict*)domain)->operator [](nextKeyPtr->c_str());
	}
	

	if (linkList.size()-1 == 0) /* Minus 1 because we haven't "popped" it off yet. I.E Shrunk the list. */
	{
		delete linkList[0];		
		return nextObject;
	}
	else
	{
		List newList = List();
		for (unsigned int i=1;i < linkList.size();i++)
		{
			void* item = linkList[i];
			newList.push_back(item);
		}

		delete linkList[0];
		return Get(newList,nextObject);
	}

}

/* Functions for reading bytes in */

short PyMarkLoader::ReadShort(ifstream* stream)
{
	char p1;
	stream->read(&p1,1);

	return (short)p1;
}

int PyMarkLoader::ReadInt(ifstream* stream)
{
	char p1;
	stream->read(&p1,1);
	char p2;
	stream->read(&p2,1);

	int retVal = ((short)p1 << 8) + ((short)p2);
	
	return retVal;
}

long PyMarkLoader::ReadLong(ifstream* stream)
{
	char p1;
	stream->read(&p1,1);
	char p2;
	stream->read(&p2,1);
	char p3;
	stream->read(&p3,1);
	char p4;
	stream->read(&p4,1);

	long retVal = ((short)p1 << 24) + ((short)p2 << 16) + ((short)p3 << 8) + ((short)p4);
	
	return retVal;
}

string PyMarkLoader::ReadString(ifstream *stream)
{
	int len = ReadInt(stream);

	string retString = "";
	for(int i=0;i<len;i++)
	{
		char p1;
		stream->read(&p1,1);
		retString += p1;
	}

	return retString;
}

string PyMarkLoader::ReadLongString(ifstream *stream)
{
	long len = ReadLong(stream);

	string retString = "";
	for(long i=0;i<len;i++)
	{
		char p1;
		stream->read(&p1,1);
		retString += p1;
	}

	return retString;
}

float PyMarkLoader::ReadFloat(ifstream *stream)
{
	char p1;
	stream->read(&p1,1);
	char p2;
	stream->read(&p2,1);
	char p3;
	stream->read(&p3,1);
	char p4;
	stream->read(&p4,1);

	char arr[4] = {p4,p3,p2,p1};
	
	float ret;
	memcpy(&ret,&arr,4);

	return ret;
}

double PyMarkLoader::ReadDouble(ifstream *stream)
{
	char p1;
	stream->read(&p1,1);
	char p2;
	stream->read(&p2,1);
	char p3;
	stream->read(&p3,1);
	char p4;
	stream->read(&p4,1);
	char p5;
	stream->read(&p5,1);
	char p6;
	stream->read(&p6,1);
	char p7;
	stream->read(&p7,1);
	char p8;
	stream->read(&p8,1);


	char arr[8] = {p8,p7,p6,p5,p4,p3,p2,p1};
	
	double ret;
	memcpy(&ret,&arr,8);

	return ret;
}